src/ui/units/datasets/components/SelectSourcePrototypes/SourcesTable.tsx

```diff
@@ -108,43 +121,126 @@ export const SourcesTable: React.FC<SourcesTableProps> = ({
     onDelete,
     onRetry,
 }) => {
-    if (loading) {
-        return (
-            <div className={b('loader-sources')}>
-                <Loader size="s" />
-            </div>
-        );
-    }
+    const [search, setSearch] = React.useState('');
+    const dispatch = useDispatch();
+    const {source_listing} = useSelector(optionsSelector);
+    const currentDbName = useSelector(currentDbNameSelector);
+    const sourcesPagination = useSelector(sourcesPaginationSelector);
+    const sourcesError = useSelector(sourcesErrorSelector);
+    const dbNames = useSelector(currentDbNamesSelector);
+
+    const serverPagination = source_listing?.supports_source_pagination;
+    const serverSearch = source_listing?.supports_source_search;
+    const canChangeSourceParent = source_listing?.supports_db_name_listing;
+    const parentLabel = source_listing?.db_name_label ?? '';
+
+    React.useEffect(() => {
+        setSearch(sourcesPagination.searchValue);
+    }, [sourcesPagination.searchValue]);
+
+    const onChangeDbName = React.useCallback(
+        (value) => {
+            dispatch(changeCurrentDbName(value[0]));
+        },
+        [dispatch],
+    );
+
+    const onLoadMore = React.useCallback(() => {
+        if (
+            !sourcesPagination.isFetchingNextPage &&
+            !loading &&
+            !sourcesPagination.isFinished &&
+            !sourcesError
+        ) {
+            dispatch(incrementSourcesPage());
+        }
+    }, [dispatch, loading, sourcesPagination, sourcesError]);
+
+    const debouncedSearch = React.useMemo(
+        () =>
+            debounce((value: string) => {
+                dispatch(searchSources(value));
+            }, SEARCH_DELAY),
+        [],
+    );
+
+    const onSearch = React.useCallback(
+        (value: string) => {
+            const currentValue = value.trim();
+            setSearch(currentValue);
+            debouncedSearch(currentValue);
+        },
+        [debouncedSearch],
+    );
 
     return (
         <div className={b('sources')}>
             {error ? <ErrorView onRetry={onRetry} error={error} /> : null}
             <div className={b('top-section')}>
+                {canChangeSourceParent && (
+                    <>
+                        <div className={b('header')}>
+                            <span>{parentLabel}</span>
+                        </div>
+                        <Select
+                            disabled={loading}
+                            className={b('select-db-name')}
+                            value={[currentDbName ? currentDbName : '']}
+                            onUpdate={onChangeDbName}
+                        >
+                            {dbNames?.map((name: string) => (
```

Comment:
According to the code, the actual type of dbNames is (string | undefined)[]. It's also worth making a more unique key here.